Implemented the geo_tweets function and added georadius_tweets function

yvan opened the pull request:
Hey I forked the repository a few days ago. Have been looking through the code. I noticed there were some functions in the documentation that were not implemented so I implemented them. I implemented the geo_tweets method and added the georadius_tweets method in the get_tweets.py file. I also edited the geo_tweet.py file and combined all the continent geoboxes into an array which can then be more easily passed around. I tested the methods I added, and they seem to work. I also edited the README to include instructions for using the new functions as well as a description of what oauth json files (containing secrets/keys) should look like in terms of format. If there's anything wrong let me know and I'll fix it.

yvan committed:
added the geo_tweets method to the get tweets file, implemented as close to the docs as I could get but currently feeding a list of locations will just output a list of all locations in the same array without any defining indexes from one country/place over another. added a small amount of code in geo_tweet.py to organize continent codes as an array for later use

yvan committed:
fixed the oAuth file format section on the docs. I added this section because it was unclear what format the tweepy API pool function expected in a json file, and the format it expected was non-obvious, adding it saves anyone who wants to use the code about 30min to 1 hour of time fiddling with formats

yvan committed:
another readme edit on file formats

yvan committed:
yet more readme edits

yvan committed:
added a method that gets tweets from geocodes or lists of geocodes. in get_tweets called geocode_tweets

yvan committed:
changed the geocode_tweet function to be called georadius_tweets, basically it can run queries within a certain geo radius

yvan committed:
got rid of some code I added in get_tweets

yvan committed:
in geo_tweets.py file simplified the function parameter regions=[] to be regions=ContinentsGeoBoxes, a little bit easier to read, and in the future if we want to get all tweets from a certain contient (using streaming API and bounding boxes) it will also use the ContinentsGeoBoxes array

yvan committed:
minor edits on continents geoboxes

yvan committed:
fixed a problem with the query input on the geo_tweets and georadius_tweets functions in the get_tweets.py file

yvan committed:
added use case for the geo_tweets and georadius_tweets to the readme

yvan committed:
change format of REAMEM to make sense in geo_tweets section

dpenfoldbrown commented:
Hey Yvan, 

Generally, looks good. Functionality is good. A few comments before I pull:

1) I haven't worked on this code in a while, and some things I'm doing are a bit archaic. For example, returning lists of tweets as opposed to iterators over them (which causes greater mem consumption, etc). This should be avoided.

2) Function names in get_tweets.py: geo_tweets should now be something more reflective of its nature, such as place_tweets.

3) In both geo_tweets and georadius_tweets, I would avoid using two different function params for a single-v-multiple option. It creates more params and duplicate code. Instead, take one param that is a list (can easily be a list of one). This also removes the need to check for either param and then throw an exception.

Also :-), there are definitely things in my code I'd improve if I had the time (or, will when I have the time).

Best,

Duncan

yvan committed:
changed name of geo_tweets to place_tweets reduced single_ and list_ variables to reduce duplicate code in georadius_tweets and place_tweets methods, return iterators instead of lists full of data, all work done in get_tweets.py

yvan committed:
fixed problem with names of imported smapp in README for place_tweets and georadius_tweets examples.

yvan committed:
added comment for geo_radius list to README

yvan commented:
# Alright I implemented all the changes you requested:

1) "change function names": Changed the name to place_tweets in the code and in the docs.

2) "fewer variables/duplicate code": My perspective was to try to make it as easy and 
obvious from the perspective of the person using the method. It didn't seem like having a 
format of [[Glasgow]] was super obvious from a user perspective. What I've done is convert 
everything from "double list" format to "single list" format, getting rid of duplicate code.

So instead of: 

[['Glasgow']] and [['Glasgow'],['London']]

we will have:

['Glasgow'] and ['Glasgow', 'London']

For the georadius_tweets function I have used the double list notation as each query needs three data points; this got rid of duplicate code:

we have:

[[37.781157,-122.398720,"1mi"]] 

and:

[[32.781830, -96.795860,"1mi"], [37.781157,-122.398720,"1mi"]]

# Cursors/iterators:

I saw the tweepy Cursor objects, and read up on tweepy docs. I agree they are super 
powerful and superior to lists in the case of the single query. They don't 
seem super conducive to dealing with the returns of multiple queries at once.

``` python
query_result = smapp.place_tweets(['Glasgow', 'London', 'York'])

#now the user can post-process query_results list
```

The code I just added returns a list with the iterator item for each query for each place like so:

[glasgow_iterator, london_iterator, york_iterator]

Do you think there's a better way to handle a method call that calls a list of multiple places or geo-radii, how should we return iterators in that scenario?

I tested the new methods, they seem to work.

Have a good weekend :-).

yvanscher commented:
Excuse the crazy formatting on that last post, I formatted it as markdown, and

dpenfoldbrown commented:
Hey,

Sounds great! In terms of merging results from multiple iterators/cursors, there is a way to do this (we do it somewhere in some other code, for example in smapp-toolkit). Let me track that down and check it out.

Otherwise, will review and pull when I get a sec.

Thanks again,

Duncan

yvan committed:
made the functions return iterators and not lists by converting the lists via iter()

yvan commented:
Hey Duncan,

it was cool to meet in person today. I made the methods return iterators of iterators instead of lists of iterators. I just converted the `list[with_iterators, inside_of_it]` to an iterator by calling `iter(list)`. That way the methods will be consistent with the rest of the code. None of the usage for my methods has changed. I tested to make sure that the methods were returning memory references to iterator objects at each stage and not entire lists of data.

I looked more at the toolkit code. I somehow missed the `_iter_` declaration at the bottom of the twitter.py file before. I'm not sure the same thing can be done for the get_tweets.py file. The `_iter_` function basically returns an instance of the `MongoTweetCollection` object and "wraps" tweets as an  iterator, just like you said. In get_tweets there is no analogous overarching class object. Anyways the code returns iterators now.

Peace,

yvan committed:
changed the name of the internal list in get_tweets.py/place_tweets to be locations_iterators and not locations_curors

jonathanronen commented on README.md:
````diff
@@ -62,12 +62,51 @@ We also use the [pip](http://www.pip-installer.org/en/latest/) package managemen
+
+```python
+
+import smappPy.get_tweets as smapp
````
Let's not suggest in the doc that users rename the module they're importing here. We don't want to encourage that people think of `get_tweets` as `smapp`, right?

jonathanronen commented on smappPy/get_tweets.py:
```diff
@@ -49,12 +46,54 @@ def user_tweets(api, user_id=None, screen_name=None, limit=None):
+        tweets_from_place = query_tweets(api, query=query+"&geocode:%s" % ",".join(georadius), limit=limit)
+        locations_iterators.append(tweets_from_place)
 
+    return iter(locations_iterators)
```
We want the API to be `for tweet in place_tweets(...):`, right?

I think that's better than

``` python
for it in place_tweets():
    for tweet in it:
```

So how about something like `return (tweet for cursor in location_iterators for tweet in cursor)`?

jonathanronen commented on smappPy/get_tweets.py:
```diff
@@ -49,12 +46,54 @@ def user_tweets(api, user_id=None, screen_name=None, limit=None):
+        tweets_from_place = query_tweets(api, query=query+"&place:%s" % place_id, limit=limit)
+        locations_iterators.append(tweets_from_place)
+
+    return iter(locations_iterators)
```
Generator here too

yvan committed:
1. fixed README place_tweets and georadius_tweets usage to no longer suggest a weird renaming in the import statement 2.fixed the methods in geo_tweets to return via generators, each method is now like two lines of code, cool

yvan commented:
I changed the functions to return via generators. I changed the README.md import statement; that was awkward.

Holy crap! Generators are cool. I can write  each of these methods in essentially two lines of code. I've never worked with this before. It only returns each tweet object one at a time when we need it. That's what makes sense for processing a huge amount of data. That will save a lot of memory.

jonathanronen commented:
Beautiful, thanks!
